Dialogs defined as a Style w/ ControlTemplate and Icons

Time2Code0 opened the pull request:
Changed the Dialogs from a UserControl to a Style with ControlTemplate (see #4168) this also opens the option of adding Icons to it as suggested in #2282 (still WIP).

- [x]  Define Dialogs as a Style
- [x] Display Icons

P.S: This is my first time working on an open source project, so let me know what I can improve/do better next time :)

Greetings

Closes #4168 
Closes #1474 
Closes #2282

punker76 commented on src/global.json:
```diff
@@ -1,6 +0,0 @@
-{
-  "sdk": {
-    "version": "5.0.301",
-    "rollForward": "latestFeature"
-  }
-}
```
Please revert this deleted file. thx

punker76 commented on src/MahApps.Metro/Controls/Dialogs/InputDialog.cs:
```diff
@@ -0,0 +1,299 @@
+﻿using System;
```
Please add these lines here, like in the other `*.cs` files :
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

```

punker76 commented on src/MahApps.Metro/Controls/Dialogs/LoginDialog.cs:
```diff
@@ -1,8 +1,4 @@
-// See the LICENSE file in the project root for more information.
-
-using System;
+﻿using System;
```
Please add these lines here, like in the other `*.cs` files :
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

```

punker76 commented on src/MahApps.Metro/Controls/Dialogs/MessageDialog.cs:
```diff
@@ -1,11 +1,8 @@
-// See the LICENSE file in the project root for more information.
-
-using System;
+﻿using System;
```
Please add these lines here, like in the other `*.cs` files :
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

```

punker76 commented on src/MahApps.Metro/Controls/Dialogs/ProgressDialog.cs:
```diff
@@ -1,9 +1,6 @@
-// See the LICENSE file in the project root for more information.
-
-using System.Threading;
+﻿using System.Threading;
```
Please add these lines here, like in the other `*.cs` files :
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

```

punker76 reviewed (commented):
Thx you for your contribution! Please look at my first comments. Your changes are looking good.

Time2Code0 commented:
> Thx you for your contribution! Please look at my first comments. Your changes are looking good.

Done. Added the gobal.json again (I deleted it on my local machine because it could not resolve the .Net SDK version and I was too lazy to install the correct one) and the missing License Text again.

Greetings

Time2Code0 commented:
Good evening,

I have added the Icon Presenter to the other dialogs. The current status is shown in the picture below. Let me know what you think about it.

Greeting

![IconDIalog](https://user-images.githubusercontent.com/47297708/142036950-60102784-a384-4c65-a6a4-99ef8258ccc8.PNG)

punker76 commented on src/MahApps.Metro/Themes/Dialogs/LoginDialog.xaml:
```diff
@@ -106,7 +112,110 @@
                         </StackPanel>
                     </Grid>
                 </Grid>
+            </Grid>-->
```
I think this can be removed.

punker76 commented on src/MahApps.Metro/Controls/Dialogs/ProgressDialog.cs:
```diff
@@ -88,6 +92,31 @@ public string NegativeButtonText
+            get => (DataTemplate?)this.GetValue(IconTemplateProperty);
+            set => this.SetValue(IconTemplateProperty, value);
+        }
+
```
Is it not better to move this to the base class? Because it is used in each derived class. So a user can use this in a custom dialog too. WDYT?

punker76 commented on src/MahApps.Metro/Controls/Dialogs/MessageDialog.cs:
```diff
@@ -131,7 +135,32 @@ public MessageDialogStyle ButtonStyle
         }
 
-        #endregion Properties
+        /// <summary>Identifies the <see cref="Icon"/> dependency property.</summary>
+        public static readonly DependencyProperty IconProperty
+            = DependencyProperty.Register(nameof(Icon),
+                                          typeof(object),
+                                          typeof(MessageDialog),
+                                          new PropertyMetadata());
+
+        public object? Icon
+        {
+            get => this.GetValue(IconProperty);
+            set => this.SetValue(IconProperty, value);
+        }
+
+        /// <summary>Identifies the <see cref="IconTemplate"/> dependency property.</summary>
+        public static readonly DependencyProperty IconTemplateProperty
+            = DependencyProperty.Register(nameof(IconTemplate),
+                                          typeof(DataTemplate),
+                                          typeof(MessageDialog));
+
+        public DataTemplate? IconTemplate
+        {
+            get => (DataTemplate?)this.GetValue(IconTemplateProperty);
+            set => this.SetValue(IconTemplateProperty, value);
+        }
```
Is it not better to move this to the base class? Because it is used in each derived class. So a user can use this in a custom dialog too. WDYT?

punker76 commented on src/MahApps.Metro/Controls/Dialogs/LoginDialog.cs:
```diff
@@ -206,6 +210,31 @@ public bool RememberCheckBoxChecked
+        {
+            get => (DataTemplate?)this.GetValue(IconTemplateProperty);
+            set => this.SetValue(IconTemplateProperty, value);
+        }
```
Is it not better to move this to the base class? Because it is used in each derived class. So a user can use this in a custom dialog too. WDYT?

punker76 commented on src/MahApps.Metro/Controls/Dialogs/InputDialog.cs:
```diff
@@ -278,21 +282,18 @@ protected override void OnLoaded()
             this.Icon = this.DialogSettings.Icon;
```
Is it not better to move this to the base class? Because it is used in each derived class. So a user can use this in a custom dialog too. WDYT?

punker76 reviewed (changes requested):
It looks really good. Can you take a look to my new comments? And the external Dialogs doesn't close anymore by using the buttons (but could be a bug before too). Thx

Time2Code0 commented:
You are absolutely correct that the Icon and IconTemplate should be a Proeprty in the BaseMetroDialog. As far as I can tell, the problem with the external Dialog is that OnApplyTemplate is executed after the WaitForButtonPressAsync method, for this reason the controls are null and the EventHandle is not applied to them.

Greetings

punker76 commented on src/MahApps.Metro.Samples/MahApps.Metro.Demo/MainWindow.xaml:
```diff
@@ -20,13 +20,14 @@
                  Icon="mahapps.metro.logo2.ico"
                  NonActiveGlowBrush="#CDFF0000"
                  ShowIconOnTitleBar="True"
+                 ShowTitleBar="{Binding ShowMyTitleBar, Mode=TwoWay}"
```
@Time2Code0 Can you remove this too?

Time2Code0 replied:
Of course. I thought I had removed it, but as you can see, I missed it.

punker76 commented:
@Time2Code0 your changes have been merged, thanks for your contribution 👍
